core/validation/specific_validations/towns_fund_round_four.py

```diff
@@ -374,43 +372,62 @@ def validate_funding_spent(workbook: dict[str, pd.DataFrame]) -> list["TownsFund
-                message=msgs.OVERSPEND_PROJECT,
-                # first project begins after 17 and each project is seperated by 28 cells
-                row_index=17 + 28 * get_project_number_by_id(project_id, project_ids),
-            )
-            for project_id in project_ids
-            if round(funding_spent[project_id], 2) > DefaultConfig.TF_FUNDING_ALLOCATED[project_id]
-        ]
-
-    return funding_spent_failures
+        funding_spent_failures = []
+        for expense_type in ["CDEL", "RDEL"]:
+            for project_id in project_ids:
+                project_number = get_project_number_by_id(project_id, project_ids)
+                if round(funding_spent[expense_type][project_id]) > get_allocated_funding(project_id, expense_type):
+                    funding_spent_failures.append(
+                        TownsFundRoundFourValidationFailure(
+                            sheet="Funding",
+                            section=f"Project Funding Profiles - Project {project_number}",
+                            column="Grand Total",
+                            message=msgs.OVERSPEND.format(expense_type=expense_type),
+                            row_index=13 + 28 * project_number if expense_type == "CDEL" else 16 + 28 * project_number,
+                        )
+                    )
+        return funding_spent_failures
 
 
-def spend_per_project(funding_df: pd.DataFrame, project_id: str) -> float:
+def spend_per_project(funding_df: pd.DataFrame, project_id: str) -> dict[str:int]:
     """return the total funding spent per an individual project
 
     :param funding_df: A dataframe of the funding table from a submission
     :param: project_id: ID of project in question
 
     :return: Total funding spent per project
     """
-    funding_spent = (
-        funding_df["Spend for Reporting Period"]
-        .loc[
-            (funding_df["Project ID"] == project_id)
-            & (funding_df["Funding Source Type"] == "Towns Fund")
-            & ~(funding_df["Funding Source Name"].str.contains(r"contractually committed"))
-        ]
-        .sum()
-    )
+    funding_df = funding_df.loc[
+        (funding_df["Project ID"] == project_id)
+        & (funding_df["Funding Source Type"] == "Towns Fund")
+        & ~(funding_df["Funding Source Name"].str.contains("contractually committed"))
+    ]
+
+    funding_spent = {
+        "index": project_id,
+        "CDEL": (
+            funding_df["Spend for Reporting Period"].loc[(funding_df["Funding Source Name"].str.contains("CDEL"))].sum()
+        ),
+        "RDEL": (
+            funding_df["Spend for Reporting Period"].loc[(funding_df["Funding Source Name"].str.contains("RDEL"))].sum()
+        ),
+        "Total": (funding_df["Spend for Reporting Period"].sum()),
+    }
     # Business logic here is taken from spreadsheet 4a - Funding Profile Z45 for grand total expenditure
     return funding_spent
 
 
+def get_allocated_funding(idx: str, expense_type: str) -> int:
```

Comment:
Does this always return an `int`?

Reply:
I've changed the function so it casts to int